### PR TITLE
Pre-construct TagMap.Entry objects in InternalTagsAdder

### PR DESCRIPTION
**Pull request by dougqh**
## What Does This Do

`InternalTagsAdder` sets `_dd.base_service` / `version` on spans via `TagMap.set(tag, value)`, which allocates a fresh `TagMap.Entry` **per span**. Both values are fixed for the life of the tracer, and `TagMap.Entry` objects are explicitly safe to share across maps (the `OptimizedTagMap` collision design via `BucketGroup` relies on it). This pre-builds the two `Entry` objects once in the constructor and reuses them via `set(entry)`.

## Motivation

A JFR allocation profile of spring-petclinic under load (full agent, 2026-06-03) attributed **~52 allocation samples to `InternalTagsAdder.processTags`** — one `Entry` allocated per span for `base.service`. Tag-map handling (`TagMap$Entry` + `OptimizedTagMap`) was the single largest non-core tracer allocation theme in that profile, ahead of the CSS metrics system. This change drops the `InternalTagsAdder` contribution to zero.

## Notes

- Re-applies the logic from #10965 (`dd/pre-construct-tagmap-entries`), which had drifted 415 commits behind master and changed `processTags`'s signature; this version sits on current master with the unchanged `AppendableSpanLinks` signature. **Supersedes #10965.**
- Behavior preserved: `Entry.stringValue()` for the OBJECT entry resolves to `obj.toString()` (cached), identical to the prior `ddService.toString()` in the `equalsIgnoreCase` comparison. The only edge case — empty `ddService` now yields an early return — is unreachable in production (`Config.getServiceName()` is never `""`).
- Existing test `InternalTagsAdderTest.groovy` passes unchanged.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java [hunk 1]
```diff
@@ -10,29 +10,38 @@
...
+  // safe to share across maps) avoids allocating a fresh Entry per span in processTags.
+  private final TagMap.Entry baseServiceEntry;
+  private final TagMap.Entry versionEntry;
 
   public InternalTagsAdder(@Nullable final String ddService, @Nullable final String version) {
-    this.ddService = ddService != null ? UTF8BytesString.create(ddService) : null;
-    this.version = version != null && !version.isEmpty() ? UTF8BytesString.create(version) : null;
+    this.baseServiceEntry =
+        ddService != null
+            ? TagMap.Entry.create(DDTags.BASE_SERVICE, UTF8BytesString.create(ddService))
```
[the remark this answers is not in the document]
Comment (dougqh) [reply dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java]: Good catch — fixed in f5f5cc90ef. The constructor now keeps a separate `ddService` `UTF8BytesString` handle that drives both the early-return guard and the case-insensitive service-name comparison (independent of whether `baseServiceEntry` was prebuilt). For the empty-`DD_SERVICE` case, `Entry.create` still returns `null`, so the base.service branch falls back to `unsafeTags.set(BASE_SERVICE, ddService)` — preserving byte-identical behavior — and crucially the `version` branch is still reached when the span's service name also matches the empty configured service.

I added regression coverage in the migrated `InternalTagsAdderTest` (now JUnit 5): `new InternalTagsAdder("", "1.0")` with an empty span service still adds version `1.0`, and with a non-empty span service it takes the base.service branch and adds no version.

(Reply by Claude)

File: dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+import java.util.stream.Stream;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class InternalTagsAdderTest {
```
Comment (amarziali) [inline dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java]: Even if we have policy to migrate to junit, did you add some new test? Otherwise it can be migrated separately

File: dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java [hunk 3]
```diff
@@ -10,12 +10,29 @@
 import javax.annotation.Nullable;
 
 public final class InternalTagsAdder extends TagsPostProcessor {
+  // ddService drives the guard and the service-name comparison; kept even when empty so behavior
+  // matches the prior implementation exactly (an explicitly-empty DD_SERVICE is a valid, if
+  // degenerate, config -- see getStringExcludingSource, which passes "" through, not the default).
```
Comment (amarziali) [inline dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java]: This looks a generated comment that is not adding information about the ddService variable itself but more related to the changes on this PR specifically

File: dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java [hunk 4]
```diff
@@ -10,12 +10,29 @@
...
   private final UTF8BytesString ddService;
-  private final UTF8BytesString version;
+
+  // base.service / version values are fixed for the life of the tracer, and TagMap.Entry objects
+  // are safe to share across maps (the OptimizedTagMap collision design relies on it), so the
+  // entries are built once and reused on the hot path instead of allocating one per span.
+  // baseServiceEntry is null when ddService is null OR empty (Entry.create rejects empty values);
+  // the empty case falls back to set(tag, value) below to preserve byte-identical behavior.
```
Comment (amarziali) [inline dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java]: That's useful I would make it more concise (looks generated)

**Review by amarziali — COMMENTED** [review]
it looks a good idea. I left improvements for the comments specifically that looks a bit too verbosely generated

File: dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java (hunk 2 above)
Comment (dougqh) [reply dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java]: Yes, I believe some tests were added to cover some corner cases.

File: dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java [hunk 5]
```diff
@@ -11,28 +11,35 @@
...
+        version != null && !version.isEmpty()
+            ? TagMap.Entry.create(VERSION, UTF8BytesString.create(version))
+            : null;
   }
 
   @Override
   public void processTags(
       TagMap unsafeTags, DDSpanContext spanContext, AppendableSpanLinks spanLinks) {
-    if (spanContext == null || ddService == null) {
+    if (spanContext == null || ddService == null || ddService.length() == 0) {
```
Comment (amarziali) [inline dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java]: when that processor is created (in `createEagerChain`) the `ddService` value is taken from `Config.getServiceName()` and this cannot be neither null neither empty. So it looks that instead adding a check for the size we can also remove the null check. I would suggest to improve the documentation (i.e. also add `@NonNull` to the constructor and simplify that instructions

File: dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java [hunk 6]
```diff
@@ -67,4 +68,16 @@ void shouldAddVersionWhenDdServiceEqualsServiceNameAndVersionSet(
     verify(spanContext, times(1)).getServiceName();
     assertEquals(expected, Objects.toString(unsafeTags.get(VERSION), null));
   }
+
+  // Regression: empty DD_SERVICE is treated the same as unset — processTags exits early and
+  // writes no tags, regardless of the span's service name.
+  @Test
+  void emptyDdServiceWritesNoTags() {
```
Comment (amarziali) [inline dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java]: This test should not be added. I don't understand why the `regression` word is used here. And the service name cannot be empty in our tracer. I would suggest to remove it and improve the documentation of this processor as suggested above

File: dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java (hunk 5 above)
Comment (dougqh) [reply dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java]: This was primarily out of an abundance of caution.  Although, both AI and human reviewers had raised this as potential concern, so it was easiest to just be cautious.

If you are confident that empty won't reach here, then we can change it.

File: dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java (hunk 6 above)
Comment (dougqh) [reply dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java]: It is a change in behavior of the `InternalTagsAdder`, so this test was guarding against that change/regression being introduced again.   Mostly, it just comes down to the contract being unclear before, so the AI had to be cautious.

File: dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java (hunk 5 above)
Comment (dougqh) [reply dd-trace-core/src/main/java/datadog/trace/core/tagprocessor/InternalTagsAdder.java]: From Claude...
Good call — confirmed that `createEagerChain()` always passes `Config.getServiceName()` which defaults to `"unnamed-java-app"` and is never null or empty. Changed the constructor param to `@Nonnull`, dropped the null check on `ddService`, and simplified the guard in `processTags` to just `spanContext == null`.

File: dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java (hunk 6 above)
Comment (dougqh) [reply dd-trace-core/src/test/java/datadog/trace/core/tagprocessor/InternalTagsAdderTest.java]: From Claude...
Agreed — removed the test and the `@Test` import. The contract is now documented by the `@Nonnull` annotation on the constructor instead.

**Review by bric3 — APPROVED** [review]

**Review by sarahchen6 — APPROVED** [review]
